rib/rib.go

```diff
@@ -1604,3 +1603,90 @@ func (r *RIBHolder) GetRIB(filter map[spb.AFTType]bool, msgCh chan *spb.GetRespo
 
 	return nil
 }
+
+type FlushErr struct {
+	Errs []error
+}
+
+func (f *FlushErr) Error() string {
+	b := &bytes.Buffer{}
+	for _, err := range f.Errs {
+		b.WriteString(fmt.Sprintf("%s\n", err))
+	}
+	return b.String()
+}
+
+// Flush cleanly removes all entries from the specified RIB. This is achieved
+// through sequentially calling DeleteXXX for the entries that are contained within
+// it. We use the Delete methods to ensure that we call all relevant hooks as changes
+// are made.
+//
+// The order of operations for deletes considers the dependency tree:
+//  - we remove IPv4 entries first, since these are never referenced counted,
+//    and the order of removing them never matters.
+//  - we check for any backup NHGs, and remove these first - since otherwise we
+//    may end up with referenced NHGs. note, we need to consider circular references
+//	  of backup NHGs, which we may allow today. we remove the backup NHGs.
+//  - we remove the remaining NHGs.
+//  - we remove the NHs.
```

Comment:
This is my lack of understanding. I would have thought we would remove NHs before NHG.

Reply:
NHGs depend on NHs -- we're essentially trying to make sure that we don't remove anything that has a non-zero ref-count, so we go and remove the NHGs to ensure that the NHs have a zero ref-count and then remove those.